lib/matplotlib/backends/backend_qt.py

```diff
@@ -682,6 +682,108 @@ def set_window_title(self, title):
         self.window.setWindowTitle(title)
 
 
+class _IconEngine(QtGui.QIconEngine):
+    """
+    Custom QIconEngine that automatically handles DPI scaling for tools icons.
+
+    This engine provides icons on-demand with proper scaling based on the current
+    device pixel ratio, eliminating the need for manual refresh when DPI changes.
+    """
+
+    def __init__(self, image_path, toolbar=None):
+        super().__init__()
+        self.image_path = image_path
+        self.toolbar = toolbar
+
+    def _is_dark_mode(self):
+        return self.toolbar.palette().color(self.toolbar.backgroundRole()).value() < 128
+
+    def paint(self, painter, rect, mode, state):
+        """Paint the icon at the requested size and state."""
+        pixmap = self.pixmap(rect.size(), mode, state)
+        if not pixmap.isNull():
+            painter.drawPixmap(rect, pixmap)
+
+    def pixmap(self, size, mode, state):
+        """Generate a pixmap for the requested size, mode, and state."""
+        if size.width() <= 0 or size.height() <= 0:
+            return QtGui.QPixmap()
+
+        # Get the current device pixel ratio
+        dpr = (self.toolbar.devicePixelRatioF() or 1) if self.toolbar else 1
```

Comment:
Idea for further simplification:

move this into a private function `_devicePixelRatio()`.

Then ...